substratevm/src/com.oracle.svm.native.libchelper/src/cpuid.c

```diff
@@ -75,106 +94,509 @@ int get_cpuid (unsigned int leaf, unsigned int *eax, unsigned int *ebx, unsigned
-#define bit_BMI2_compat          0x00000100
-#define bit_AVX512F_compat       0x00010000
-#define bit_AVX512DQ_compat      0x00020000
-#define bit_AVX512PF_compat      0x04000000
-#define bit_AVX512ER_compat      0x08000000
-#define bit_AVX512CD_compat      0x10000000
-#define bit_AVX512BW_compat      0x40000000
-#define bit_AVX512VL_compat      0x80000000
-#define bit_RTM_compat           0x00000800
-#define bit_ADX_compat           0x00080000
-#define bit_SSE4a_compat         0x00000040
-#define bit_LZCNT_compat         0x00000020
-#define bit_HTT_compat           0x10000000
-#define bit_CMOV_compat          0x00008000
-#define bit_FXSAVE_compat        0x01000000
-#define bit_MMX_compat           0x00800000
-#define bit_SSE_compat           0x02000000
-#define bit_SSE2_compat          0x04000000
-#define bit_SSE3_compat          0x00000001
-#define bit_SSSE3_compat         0x00000200
-#define bit_POPCNT_compat        0x00800000
-#define bit_AVX_compat           0x10000000
-#define bit_SHA_compat           0x20000000
-#define bit_FMA_compat           0x00001000
-
 /*
- * Extracts the CPU features by using cpuid.h.
- * Note: This function is implemented in C as cpuid.h
- * uses assembly and pollutes registers; it would be
- * difficult to keep track of this in Java.
- */
-void determineCPUFeatures(CPUFeatures* features) {
-  unsigned int eax, ebx, ecx, edx;
-  unsigned int max_level, ext_level;
-  unsigned int vendor;
+// cpuid information block.  All info derived from executing cpuid with
+// various function numbers is stored here.  Intel and AMD info is
+// merged in this block: accessor methods disentangle it.
+typedef struct {
+    // cpuid function 0
+    uint32_t std_max_function;
+    uint32_t std_vendor_name_0;
+    uint32_t std_vendor_name_1;
+    uint32_t std_vendor_name_2;
 
-  max_level = get_cpuid_max (0, &vendor);
-  if (max_level < 1) {
-    return;
+    // cpuid function 1
+    StdCpuid1Eax std_cpuid1_eax;
+    StdCpuid1Ebx std_cpuid1_ebx;
+    StdCpuid1Ecx std_cpuid1_ecx;
+    StdCpuid1Edx std_cpuid1_edx;
+
+    // cpuid function 4 (deterministic cache parameters)
+    DcpCpuid4Eax dcp_cpuid4_eax;
+
+    // cpuid function 7 (structured extended features)
+
+    SefCpuid7Ebx sef_cpuid7_ebx;
+    SefCpuid7Ecx sef_cpuid7_ecx;
+    SefCpuid7Edx sef_cpuid7_edx;
+
+    // cpuid function 0xB (processor topology)
+    // ecx = 0
+    uint32_t tpl_cpuidB0_eax;
+    TplCpuidBEbx tpl_cpuidB0_ebx;
+
+    // ecx = 1
+    TplCpuidBEbx tpl_cpuidB1_ebx;
+
+    // cpuid function 0x80000001
+    ExtCpuid1Ecx ext_cpuid1_ecx;
+    ExtCpuid1Edx ext_cpuid1_edx;
+
+    // cpuid function 0x80000007
+    ExtCpuid7Edx ext_cpuid7_edx; // tscinv
+
+    // cpuid function 0x80000008
+    ExtCpuid8Ecx ext_cpuid8_ecx;
+
+    // cpuid function 0x8000001E // AMD 17h
+    ExtCpuid1EEbx ext_cpuid1E_ebx; // threads per core (AMD17h)
+
+    // extended control register XCR0 (the XFEATURE_ENABLED_MASK register)
+    XemXcr0Eax   xem_xcr0_eax;
+
+} CpuidInfo;
```

Comment:
This struct contains only the fields we need - i'll remove it before merging; however, I thought it was best to go ahead and fill in more features, as that likely would have to be done in the future

Reply:
Delete